src/lib/y2storage/proposal/initial_strategies/ng.rb

```diff
@@ -55,56 +56,86 @@ def initial_proposal(settings: nil, devicegraph: nil, disk_analyzer: nil)
+        def try_without_adjust_by_ram(proposal, volume, devicegraph, disk_analyzer)
           if proposal.failed? && adjust_by_ram_active_and_configurable?(volume)
-            volume.adjust_by_ram = false
-            log.info("Trying proposal after disabling 'adjust_by_ram' for '#{volume.mount_point}'")
-            proposal = try_proposal(Yast.deep_copy(current_settings), devicegraph, disk_analyzer)
+            proposal = retry_proposal(proposal, volume, :adjust_by_ram, devicegraph, disk_analyzer)
           end
           proposal
         end
 
         # Try proposal again after disabling 'snapshots'
-        def try_without_snapshots(proposal, current_settings, volume, devicegraph, disk_analyzer)
+        def try_without_snapshots(proposal, volume, devicegraph, disk_analyzer)
           if proposal.failed? && snapshots_active_and_configurable?(volume)
-            volume.snapshots = false
-            log.info("Trying proposal after disabling 'snapshots' for '#{volume.mount_point}'")
-            proposal = try_proposal(Yast.deep_copy(current_settings), devicegraph, disk_analyzer)
+            proposal = retry_proposal(proposal, volume, :snapshots, devicegraph, disk_analyzer)
           end
           proposal
         end
 
         # Try proposal again after disabling the volume
-        def try_without_proposed(proposal, current_settings, volume, devicegraph, disk_analyzer)
+        def try_without_proposed(proposal, volume, devicegraph, disk_analyzer)
           if proposal.failed? && proposed_active_and_configurable?(volume)
-            volume.proposed = false
-            log.info("Trying proposal after disabling '#{volume.mount_point}'")
-            proposal = try_proposal(Yast.deep_copy(current_settings), devicegraph, disk_analyzer)
+            proposal = retry_proposal(proposal, volume, :proposed, devicegraph, disk_analyzer)
           end
           proposal
         end
 
+        # Tries a new proposal based on the previous one after having adapted
+        # one of the volumes of the settings
+        #
+        # @note This reuses and modifies the settings from the previous
+        #   proposal, that should be fine since the new proposal is aimed to
+        #   replace the previous one.
+        #
+        # @param proposal [GuidedProposal] initial proposal, its settings will
+        #   be adapted and reused (i.e. they will be modified)
+        # @param volume [VolumeSpecification] it must be one of the volumes from
+        #   the settings of the initial proposal
+        # @param attr [Symbol] name of the attribute to disable in the volume
+        # @param devicegraph [Devicegraph] see {#try_proposal}
+        # @param disk_analyzer [DiskAnalyzer] see {#try_proposal}
+        def retry_proposal(proposal, volume, attr, devicegraph, disk_analyzer)
+          volume.send(:"#{attr}=", false)
+          adjustment = proposal.auto_settings_adjustment.add_volume_attr(volume, attr, false)
```

Comment:
NP: style note. Having three params and one of it boolean is really bad design decision. What false stands for? is it value for attr? why it is add?

For me more nature API would be `proposal.auto_settings_adjustment.attrs_for(volume)[attr] = false`

Reply:
Well, strictly speaking from the POV from which the API is designed, the last argument is not necessarily a boolean. Although right now we only disable attributes, the method is meant to assign any value that makes sense for each attribute (attributes can be boolean, strings, arrays, DiskSize objects...).

Still you suggestion would apply, I know. But I'm not sure if I fully like it... specially for the kind of partial implementation we wanted to do at this point in time (so far we only allow settings attrs to false to keep stuff deadly simple in GMC phase).